### PR TITLE
Add check for input shape

### PR DESCRIPTION
**Pull request by krishung5**
Should fix the issue #3119

### CODE REVIEW

File: src/servers/http_server.cc [hunk 1]
```diff
@@ -801,6 +801,23 @@ ValidateOutputParameter(triton::common::TritonJson::Value& io)
   return nullptr;  // success
 }
 
+TRITONSERVER_Error*
+CheckInputShape(
```
Comment (CoderHam) [inline src/servers/http_server.cc:805]: You are checking only if the total number of elements is the same and not if each dimension is the same. Maybe rename to CheckInputElementCount or something that more clearly explains what this function is doing.

File: src/servers/http_server.cc [hunk 2]
```diff
@@ -801,6 +801,23 @@ ValidateOutputParameter(triton::common::TritonJson::Value& io)
...
+    const std::vector<int64_t>& shape,
+    triton::common::TritonJson::Value& tensor_data)
+{
+  int shape_product = 1;
+  for (size_t i = 0; i < shape.size(); i++) {
+    shape_product *= shape[i];
+  }
+  if (shape_product != static_cast<int>(tensor_data.ArraySize())) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL, "Shape size is mismatched with data size");
```
Comment (CoderHam) [inline src/servers/http_server.cc:815]: Shape does not match true shape of "data" field.

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/servers/http_server.cc [hunk 3]
```diff
@@ -801,6 +801,24 @@ ValidateOutputParameter(triton::common::TritonJson::Value& io)
...
+TRITONSERVER_Error*
+CheckInputElementCount(
+    const std::vector<int64_t>& shape,
+    triton::common::TritonJson::Value& tensor_data)
+{
+  int shape_product = 1;
+  for (size_t i = 0; i < shape.size(); i++) {
+    shape_product *= shape[i];
+  }
+  if (shape_product != static_cast<int>(tensor_data.ArraySize())) {
```
Comment (NonStatic2014) [inline src/servers/http_server.cc:813]: Convert it to `int` may loss precision?

File: src/servers/http_server.cc [hunk 4]
```diff
@@ -801,6 +801,24 @@ ValidateOutputParameter(triton::common::TritonJson::Value& io)
...
 }
 
+TRITONSERVER_Error*
+CheckInputElementCount(
+    const std::vector<int64_t>& shape,
+    triton::common::TritonJson::Value& tensor_data)
+{
+  int shape_product = 1;
+  for (size_t i = 0; i < shape.size(); i++) {
+    shape_product *= shape[i];
```
[the remark this answers is not in the document]
Comment (NonStatic2014) [reply src/servers/http_server.cc:811]: Can you add UTs?

**Review by NonStatic2014 — CHANGES_REQUESTED** [review]

File: src/servers/http_server.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (GuanLuo) [reply src/servers/http_server.cc:811]: Using int64_t should be safe in terms of overflowing

File: src/servers/http_server.cc (hunk 3 above)
Comment (GuanLuo) [reply src/servers/http_server.cc:813]: Note that [TensorData](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#tensor-data) can be represent in two ways, so the function should check which representation it uses and compare the dimensions accordingly:

- if 1-D, then get element count of the shape and compare the value
- if > 1-D, then expect the number of dimensions is the same and iterate each dimension to ensure they have the same shape.

**Review by GuanLuo — CHANGES_REQUESTED** [review]
Is the issue due to the content in "data" field misaligns with the "shape" field of the input? If so, it seems to be the same issue that this [FIXME](https://github.com/triton-inference-server/server/blob/afa2cd415bfdf35f21d3eab219a31e7a9ad4bf70/src/servers/http_server.cc#L2026-L2031) refers to. I think it is better to have a explicit check instead of what is suggested in the FIXME, but you should clean up the FIXME and related section as your change will address them.

File: src/servers/http_server.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (NonStatic2014) [reply src/servers/http_server.cc:811]: but i will still suggest adding UTs to prevent regressions later.

File: src/servers/http_server.cc (hunk 3 above)
Comment (CoderHam) [reply src/servers/http_server.cc:813]: Yes. Good catch @GuanLuo!

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/servers/http_server.cc [hunk 5]
```diff
@@ -379,6 +381,13 @@ ReadDataFromJsonHelper(
         default:
           break;
       }
+      // Check if writing to 'serialized' is overrunning the expected byte_size
+      uint64_t bytes_read = *counter * TRITONSERVER_DataTypeByteSize(dtype);
```
Comment (GuanLuo) [inline src/servers/http_server.cc:385]: You don't need to compute the byte size on every element, in fact, you can just keep track of the expected total counter (expected number of elements) and compare with `*counter` directly, as `byte_size` is simply computed by `element_cnt * TRITONSERVER_DataTypeByteSize(dtype)`

From this perspective, note that the expected counter is slightly different from `TRITONSERVER_TYPE_BYTES`, which is just the byte size, so you need to condition to set it properly before calling into `ReadDataFromJson()`

**Review by GuanLuo — CHANGES_REQUESTED** [review]

File: src/servers/http_server.cc (hunk 5 above)
Comment (krishung5) [reply src/servers/http_server.cc]: I see. Script is modified, please review.

File: src/servers/http_server.cc [hunk 6]
```diff
@@ -269,8 +270,15 @@ ReadDataFromJsonHelper(
     TRITONSERVER_Error* assert_err =
         el.AssertType(triton::common::TritonJson::ValueType::ARRAY);
     if (assert_err == nullptr) {
-      RETURN_IF_ERR(ReadDataFromJsonHelper(base, dtype, el, counter));
+      RETURN_IF_ERR(
+          ReadDataFromJsonHelper(base, dtype, el, counter, expected_cnt));
     } else {
+      // Check if writing to 'serialized' is overrunning the expected byte_size
+      if (*counter == expected_cnt) {
```
Comment (GuanLuo) [inline src/servers/http_server.cc:277]: Use >=, counter for `TRITONSERVER_TYPE_BYTES` can increment more than 1 value. You probably need to add a similar check after line 380 after you know how much more bytes are needed to store the value (i.e. need `*counter + len + sizeof(uint32_t) <= expected_cnt`)

**Review by GuanLuo — CHANGES_REQUESTED** [review]
Otherwise looks good except for one more change. And @NonStatic2014 raised a point that we should add test coverage, testing with invalid request JSON, around it. You can add it to this PR or open a follow-up PR for that

File: src/servers/http_server.cc (hunk 6 above)
Comment (krishung5) [reply src/servers/http_server.cc]: Modified. Test cases for invalid request JSON are added. Please reveiw.

**Review by GuanLuo — APPROVED** [review]

File: qa/L0_http/test.sh [hunk 7]
```diff
@@ -367,6 +367,34 @@ if [ `grep -c "\[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0\]" ./curl.out` != "1" ]; then
     RET=1
 fi
 
+# send bad request with invalid json
```
Comment (CoderHam) [inline qa/L0_http/test.sh:370]: The json itself is not invalid. Fix:
Send bad request where the 'data' field misaligns with the 'shape' field of the input

File: qa/L0_http/test.sh [hunk 8]
```diff
@@ -367,6 +367,34 @@ if [ `grep -c "\[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0\]" ./curl.out` != "1" ]; then
...
+set -e
+if [ "$code" != "400" ]; then
+    cat ./curl.out
+    echo -e "\n***\n*** Test Failed\n***"
+    RET=1
+fi
+if [ `grep -c "\{\"error\":\"Unable to parse 'data': Shape does not match true shape of 'data' field\"\}" ./curl.out` != "1" ]; then
+    RET=1
+fi
+
```
Comment (CoderHam) [inline qa/L0_http/test.sh:397]: Add a check here to ensure the sever is running and can successful infer even after the above bad request.

File: src/servers/http_server.cc [hunk 9]
```diff
@@ -411,10 +425,19 @@ ReadDataFromJson(
 
     default:
       RETURN_MSG_IF_ERR(
-          ReadDataFromJsonHelper(base, dtype, tensor_data, &counter),
+          ReadDataFromJsonHelper(
+              base, dtype, tensor_data, &counter, expected_cnt),
           "Unable to parse 'data'");
       break;
   }
+  // Check if 'ReadDataFromJsonHelper' reads
```
Comment (CoderHam) [inline src/servers/http_server.cc:433]: Run formatter

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/servers/http_server.cc (hunk 9 above)
Comment (CoderHam) [reply src/servers/http_server.cc:433]: Also add empty line before this comment

Comment (krishung5) [reply src/servers/http_server.cc]: Updated

File: qa/L0_http/test.sh [hunk 10]
```diff
@@ -394,6 +393,22 @@ fi
 if [ `grep -c "\{\"error\":\"Unable to parse 'data': Shape does not match true shape of 'data' field\"\}" ./curl.out` != "1" ]; then
     RET=1
 fi
+# Check if the server is still working after the above bad requests
```
Comment (CoderHam) [inline qa/L0_http/test.sh:396]: Add empty line before comment here

**Review by CoderHam — CHANGES_REQUESTED** [review]
Minor formatting else LGTM

File: qa/L0_http/test.sh (hunk 8 above)
Comment (CoderHam) [reply qa/L0_http/test.sh:412]: The check should be an additional successful inference -> not just checking the output from the current inference.

**Review by CoderHam — CHANGES_REQUESTED** [review]

**Review by CoderHam — APPROVED** [review]

**Pull request merged**